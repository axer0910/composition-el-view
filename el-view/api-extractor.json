{
  "extends": "../api-extractor.json",
  "mainEntryPointFilePath": "./dist/el-view-form/index.d.ts",
  "dtsRollup": {
    "untrimmedFilePath": "./dist/el-view-form.d.ts"
  }
}
