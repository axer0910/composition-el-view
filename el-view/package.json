{
  "name": "composition-el-view",
  "version": "0.1.0",
  "description": "composition-el-view",
  "main": "index.js",
  "module": "dist/composition-el-view-form.esm-bundler.js",
  "unpkg": "dist/composition-el-view-form.global.js",
  "files": [
    "index.js",
    "dist"
  ],
  "types": "dist/composition-el-view-form.d.ts",
  "buildOptions": {
    "priority": 1,
    "name": "compositionElViewForm",
    "formats": [
      "esm-bundler",
      "cjs",
      "global"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/axer0910/composition-el-view"
  },
  "keywords": [
    "vue",
    "composition-api",
    "vue-composable",
    "composable"
  ],
  "author": "axer0910",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/axer0910/composition-el-view/issues"
  },
  "sideEffects": false,
  "peerDependencies": {
    "@vue/composition-api": "^0.5.0",
    "vue": "^2.6.10",
    "element-ui": "^2.13.0"
  }
}
