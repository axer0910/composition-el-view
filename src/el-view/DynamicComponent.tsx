import { createElement } from '@vue/composition-api';
import { ComponentOptions, VNode, VNodeChildren, VNodeData } from 'vue';
import { Vue } from 'vue/types/vue';

type DynamicComponentChildren = VNodeChildren | DynamicComponentOption | DynamicComponentOption[];

export interface DynamicComponentOption {
  tag: string | VNode;
  options?: VNodeData,
  children?: DynamicComponentChildren
}

interface DynamicComponentProp {
  option: DynamicComponentOption
}

const h = createElement;

const createDynamicComponent = (dynamicComponentOption: DynamicComponentOption) => {
  const { tag, options, children } = dynamicComponentOption;
  return createElement(tag as string, options, children as VNodeChildren);
};

/*
动态创建组件：// todo 感觉这个组件意义不大，后面会删除，直接在DynamicForm组件调用createElement创建组件
tag：既可以传入字符串动态生成，也可以直接传入VNode节点
options：对应生成组件需要传入的props
children：动态组件子节点的内容，可以是普通字符串，jsx数组或节点，或者动态配置对象嵌套
 */
export default {
  props: {
    option: Object
  },
  setup: (props: DynamicComponentProp) => {
    return () => createDynamicComponent(props.option);
  }
}
