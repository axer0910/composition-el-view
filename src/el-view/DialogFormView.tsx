import { defineComponent, onMounted, Ref, ref } from '@vue/composition-api';
import { CreateElement } from 'vue';
import { ComponentRenderProxy } from '@vue/composition-api/dist/component/component';
import { FormViewState } from '@/el-view/FormState';
import { FormView, FormViewComponentContext } from '@/el-view/FormView';

export type DialogFormViewRenderContext = any;
type DialogFormViewProps = any;

export const DialogFormView = defineComponent({
  name: 'DialogFormView',components: {
    'form-view': FormView
  },
  props: {
    dialogTitle: {
      type: String,
      default: '编辑'
    },
    dialogWidth: String,
    dialogClassName: String,
    formState: Object
  },
  setup: (props) => {
    const dialogVisible = ref(false);
    const showDialog = async () => {
      dialogVisible.value = true;
      // 清理表单验证
      setTimeout(() => {
        props.formState!.getElFormRef().resetFields();
      });
    };
    const hideDialog = () => {
      dialogVisible.value = false;
    };
    const submitPromise: Ref<null | Promise<void>> = ref(null);
    const submitFn = async () => {
      if (submitPromise.value) return;
      try {
        console.log('before run submit');
        submitPromise.value = formSubmit.value();
        await submitPromise.value;
        console.log('after run submit');
        hideDialog();
      } catch (e) {
        // 验证异常
        console.log('form dialog validate error', e);
      }
      submitPromise.value = null;
    };
    const formViewRef: Ref<any> = ref(null);
    const formSubmit: Ref<() => Promise<void>> = ref(() => {});
    onMounted(async () => {
      setTimeout(() => {
        formSubmit.value = formViewRef.value!.operations.submitForm;
      })
    });
    return {
      dialogVisible, showDialog, formViewRef, submitFn, hideDialog, submitPromise
    }
  },
  render(h: CreateElement) {
    // todo 右上角小叉处理
    const renderContext = this as any;
    const props = renderContext.$props as any;
    return (
      <el-dialog title={ props!.dialogTitle }
                 visible={ renderContext.dialogVisible }
                 width={ props!.dialogWidth }
                 showClose={ true }
                 className={ props!.dialogClassName }>
        <form-view formState={ props!.formState } ref="formViewRef" showOperations={false}  v-loading={ renderContext.submitPromise !== null} />
        <span slot="footer" class="dialog-footer">
          <el-button onClick={ () => { renderContext.hideDialog() } }>取 消</el-button>
          <el-button type="primary" onClick={ () => { renderContext.submitFn() } }>确 定</el-button>
        </span>
      </el-dialog>
    );
  }
});
