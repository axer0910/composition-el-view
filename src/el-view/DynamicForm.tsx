import { CreateElement, VNode, VNodeChildren, VNodeData } from 'vue';
import { computed, createElement, onMounted, Ref, ref, SetupContext } from '@vue/composition-api';
import { getObjectValue, setObjectValue } from '@/el-view/utils';
import { defineComponent } from "@vue/composition-api";
import { Vue } from 'vue/types/vue';

export type FormOption = NullableFormItem[] | NullableFormItem[][];

interface DynamicFormProps {
  formOption: FormOption,
  formModel: {[K in keyof object]: any},
  labelWidth?: String,
  formProps?: {[key: string]: any},
  formEvents?: {[key: string]: Function},
}

export interface FormItem {
  formLabel: string,
  tagName: string,
  modelKey: string
  required?: boolean,
  requiredMsg?: string,
  formRules?: Array<ValidateRule>,
  props?: { [key: string]: any },
  attrs?: { [key: string]: any },
  events?: { [key: string]: Function },
  options?: ElOptionTag[] | VNodeChildren, // options可以传jsx，支持多个node
  className?: string
}

export type NullableFormItem = FormItem | null;

// element ui 下拉框对象快捷格式
export interface ElOptionTag {
  label: string,
  value: string | number
}

export interface ValidateRule {
  validator: (rule: any, value: any, callback: (fn?: Error) => {}) => void
  required?: boolean
}

// element ui 下拉框对象格式
export interface ElOptionTag {
  label: string,
  value: string | number
}

export interface ElTagHandler { // 注册自定义的el-tag默认属性
  defaultProps?: { [key: string]: any }, // 组件默认props
  valueChangeEventName?: string // 表单项组件触发值更新的事件名称（不传默认key为input）
  defaultBindValuePropKey?: string // 表单项组件默认绑定值的key（不传默认key为value）
  getBindValue?: (formItem: FormItem) => { valuePropKey: string, value: any }, // 获取表单项组件的绑定值
  formItemEvents?: (formItem: FormItem, changeModelEmitter: (val: any) => void) => { [key: string]: Function }, // 表单项事件处理函数
  slotBuilderFn?: (options: ElOptionTag[], formItem: FormItem) => CompVNodeData[] // 将label->value的形式的配置转换slot（el-select等场景使用）
}

export interface CompVNodeData {
  tagName: string,
  componentOption?: VNodeData,
  children?: VNodeChildren
}

const elTagHandlers: Map<string, ElTagHandler> = new Map<string, ElTagHandler>();

export const registerElTagHandlers = (tagName: string, elTagHandler: ElTagHandler) => {
  elTagHandlers.set(tagName, elTagHandler);
};

const h = createElement;
let context: SetupContext;
let componentProps: DynamicFormProps;

const emitFormChange = (val: any, modelKey: string) => {
  const { formModel } = componentProps;
  setObjectValue(formModel, modelKey, val);
  context.emit('formChange', { ...formModel });
};

// 取得绑定的prop（绑定值与其他表单项组件默认props）
const getBindProps = (formItem: FormItem) => {
  const { formModel } = componentProps;
  let props: {[key: string]: any} = {};
  if (elTagHandlers.has(formItem.tagName)) {
    // 设置定义了绑定value的prop key
    const formTagItemConf = elTagHandlers.get(formItem.tagName)!;
    const bindValuePropKey = formTagItemConf.defaultBindValuePropKey;
    if (bindValuePropKey) {
      props[bindValuePropKey] = getObjectValue(formModel, formItem.modelKey);
    } else {
      props.value = getObjectValue(formModel, formItem.modelKey)
    }
    const bindValFn = formTagItemConf.getBindValue;
    // 自定义了获取绑定值的方法
    if (bindValFn) {
      const { valuePropKey, value } = bindValFn(formItem);
      props[valuePropKey] = value;
    }
    // 设置组件默认props配置
    const { defaultProps } = formTagItemConf;
    if (defaultProps) {
      props = {
        ...props,
        ...defaultProps
      };
    }
    return props;
  } else {
    return {
      value: getObjectValue(formModel, formItem.modelKey)
    };
  }

};

// 绑定更新表单后的数据，并且触发表单更新事件
const getBindEvents = (formItem: FormItem) => {
  let events: {[key: string]: Function} = {};
  if (elTagHandlers.has(formItem.tagName)) {
    // 获取用户自定义的触发model更新的事件key
    const customDefaultInputEventName = elTagHandlers.get(formItem.tagName)!.valueChangeEventName;
    let changeEventKey;
    if (customDefaultInputEventName) {
      changeEventKey = customDefaultInputEventName;
    } else {
      changeEventKey = 'input';
    }
    events[changeEventKey] = (val: any) => {
      emitFormChange(val, formItem.modelKey);
    };
    // 获取用户自定义绑定的事件对象（emitter为触发绑定的表单model更新函数）
    const emitter = (val: any) => {
      emitFormChange(val, formItem.modelKey);
    };
    const tagEventsHandler = elTagHandlers.get(formItem.tagName)!.formItemEvents;
    if (tagEventsHandler) {
      events = {
        ...events,
        ...tagEventsHandler(formItem, emitter)
      }
    }
    return events;
  } else {
    return {
      input: (val: any) => {
        emitFormChange(val, formItem.modelKey);
      }
    };
  }
};

// 合并事件
const mergeEvents = (componentOption: VNodeData, formItem: FormItem) => {
  componentOption.on = {
    ...getBindEvents(formItem), // 获取绑定事件
    ...componentOption.on
  };
};

// 合并props
const mergeProps = (componentOption: VNodeData, formItem: FormItem) => {
  componentOption.props = {
    ...getBindProps(formItem),// 获取绑定的props（value以及其他默认props）
    ...componentOption.props
  };
};

// 生成表单项组件的slot内容
const setComponentSlot: (componentOption: VNodeData, formItem: FormItem) => VNode[] = (componentOption, formItem) => {
  const { options } = formItem;
  if (!options) return [];
  if (elTagHandlers.has(formItem.tagName)) {
    const { slotBuilderFn } = elTagHandlers.get(formItem.tagName)!;
    if (options && Array.isArray(options) && options.length > 0 && slotBuilderFn && typeof slotBuilderFn === 'function') {
      // 区分option对象格式，
      if (Reflect.has(options[0] as object, 'label')) {
        const slotData = slotBuilderFn(options as ElOptionTag[], formItem); // label，value的快捷定义
        return slotData.map(compVNodeData => createFormItemVNode(compVNodeData));
      }
    }
  }
  // options此时为VNode数组，直接返回渲染
  return options as VNode[];
};

const createFormItem = (formItem: FormItem) => {
  const { tagName, props, attrs, events, modelKey, formLabel } = formItem;
  const componentTagName = tagName;
  const componentOption: VNodeData = {
    key: modelKey,
    props,
    attrs,
    on: events
  };
  mergeEvents(componentOption, formItem);
  mergeProps(componentOption, formItem);
  const slots = setComponentSlot(componentOption, formItem);
  return (
    <el-form-item label={formLabel} class={['dy-form-item', formItem.className]} prop={formItem.modelKey}>
      {createFormItemVNode({
        tagName: componentTagName,
        componentOption,
        children: slots
      })}
    </el-form-item>
  );
};

// 如果是一维数组，那么以一列展示表单
const normalizeForm = (formOptions: FormItem[]) => {
  return formOptions.map((formItem) => [formItem]);
};

const filterNullItem = (formItem: FormItem[][]) => {
  return formItem.map(formRow => {
    return formRow.filter(formItem => formItem !== null);
  })
};

const createFormItemVNode = (compVNodeData: CompVNodeData) => {
  return createElement(compVNodeData.tagName, compVNodeData.componentOption, compVNodeData.children);
};

export const DynamicForm = defineComponent({
  props: {
    formOption: Array,
    formModel: Object,
    labelWidth: String,
    formProps: Object,
    formEvents: Object
  },
  setup: (props: DynamicFormProps, setupContext: SetupContext) => {
    context = setupContext;
    componentProps = props;
    const elFormRef: Ref<null | Vue> = ref(null);
    const _form = computed(() => {
      if (props.formOption.length > 0 && !Array.isArray(props.formOption[0])) {
        return filterNullItem(normalizeForm(props.formOption as FormItem[]));
      }
      return filterNullItem(props.formOption as FormItem[][]);
    });

    const formRules = computed(() => {
      const rules: {[key: string]: object[]} = {};
      for (const row of _form.value) {
        for (const formItem of row) {
          const defaultRequiredMsg = formItem.requiredMsg ? formItem.requiredMsg : `请填写${formItem.formLabel}`;
          if (formItem.formRules) {
            rules[formItem.modelKey] = [ ...formItem.formRules ];
          } else if (formItem.required) {
            rules[formItem.modelKey] = [{ required: true, message: defaultRequiredMsg, trigger: 'blur' }];
          } else {
            rules[formItem.modelKey] = [];
          }
        }
      }
      return rules;
    });

    onMounted(() => {
      elFormRef.value = setupContext.refs.el_form as Vue;
    });
    return {
      _form, elFormRef, formRules
    }
  },
  render(h: CreateElement) {
    const contextData = (this as typeof DynamicForm.data & typeof DynamicForm.props)!;
    const extProps = contextData.formProps ? { ...contextData.formProps } : {};
    const extEvents = contextData.formEvents ? { ...contextData.formEvents } : {};
    const formProps = {
      props: {
        rules: contextData.formRules,
        model: contextData.formModel,
        'label-width': contextData.labelWidth,
        ...extProps
      },
      on: { ...extEvents }
    };
    return (
      <el-form ref="el_form" { ...formProps }>
        {
          contextData._form.map((formOptionRow) => (
            <div class="form-row">
              {
                formOptionRow.map(formItem => (
                  <div class="form-col">
                    {createFormItem(formItem as FormItem)}
                  </div>
                ))
              }
            </div>
          ))
        }
      </el-form>
    )
  }
});
