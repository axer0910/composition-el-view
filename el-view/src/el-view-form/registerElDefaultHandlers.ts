// 注册默认的一些element ui配置
import { ElOptionTag, registerElTagHandlers } from './DynamicForm';

export const registerDefaultHandlers = () => {
  registerElTagHandlers('el-select', {
    slotBuilderFn: (options: ElOptionTag[]) => {
      return options.map(option => {
        return {
          tagName: 'el-option',
          componentOption: {
            props: {
              key: option.value,
              value: option.value,
              label: option.label
            }
          }
        };
      })
    }
  });
};
