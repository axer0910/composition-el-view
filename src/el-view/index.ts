import { DynamicForm } from '@/el-view/DynamicForm';
import { FormView } from '@/el-view/FormView';
export { useFromState } from '@/el-view/FormState';
import * as VueCompositionApi from '@vue/composition-api';

const Components = [
  DynamicForm,
  FormView
];

export const compositionApi = VueCompositionApi;

export default Components;