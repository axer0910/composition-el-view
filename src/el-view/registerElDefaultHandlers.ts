// 注册默认的一些element ui配置
import { ElOptionTag, registerElTagHandlers } from '@/el-view/DynamicForm';

export const registerHandlers = () => {
  registerElTagHandlers('el-select', {
    slotBuilderFn: (options: Array<ElOptionTag>) => {
      return options.map(option => {
        return {
          tagName: 'el-option',
          componentOption: {
            props: {
              key: option.value,
              value: option.value,
              label: option.label
            }
          }
        };
      })
    }
  });
};
