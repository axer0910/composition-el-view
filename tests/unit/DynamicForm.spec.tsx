import { createLocalVue, shallowMount, VueClass } from '@vue/test-utils';
import { FormItem, DynamicForm } from '@/el-view/DynamicForm';
import VueCompositionApi, { reactive, Ref, ref } from '@vue/composition-api';
import ElementUI from 'element-ui';
import { registerHandlers } from '@/el-view/registerElDefaultHandlers';
import Vue from 'vue';

describe('DynamicForm.tsx', () => {
  registerHandlers();
  const localVue = createLocalVue();
  localVue.use(VueCompositionApi);
  localVue.use(ElementUI, {
    size: 'small'
  });
  let formOption: Array<FormItem> = [
    {
      formLabel: '测试表单项',
      tagName: 'el-input',
      modelKey: 'test-val.test2',
      props: {
        type: 'text'
      },
      attrs: {
        placeholder: '测试表单项'
      }
    },
    {
      formLabel: '测试下拉',
      tagName: 'el-select',
      modelKey: 'test-val.test4',
      attrs: {
        placeholder: '测试表单项'
      },
      options: [
        {
          label: '下拉1',
          value: 1
        },
        {
          label: '下拉2',
          value: 2
        }
      ]
    }
  ];
  const formModel = {};
  const wrapper = shallowMount<Vue>(DynamicForm, {
    localVue,
    propsData: {
      formOption,
      formModel: ref(reactive({})).value
    }
  });
  it('should render Dynamic Form', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should update Dynamic Form when prop change', async (done) => {
    formOption = [
      {
        formLabel: '新的表单项',
        tagName: 'el-input',
        modelKey: 'test-val.test2',
        attrs: {
          placeholder: '新的表单项'
        }
      }
    ];
    wrapper.setProps({
      formOption,
      formModel
    });
    await Vue.nextTick();
    expect(wrapper).toMatchSnapshot();
    done();
  });
  it('should render jsx', async (done) => {
    const h = (wrapper.vm).$createElement;
    formOption = [
      {
        formLabel: 'jsx下拉选项',
        tagName: 'el-select',
        modelKey: 'test-val.test2',
        options: [
          <option value="1">test1</option>,
          <option value="2">test2</option>
        ]
      }
    ];
    wrapper.setProps({
      formOption
    });
    await Vue.nextTick();
    expect(wrapper).toMatchSnapshot();
    done();
  });
  it('should have default emit event', async (done) => {
    const elSelectStubVm = wrapper.findAll('el-select-stub').at(0).vm;
    expect(Reflect.has(elSelectStubVm.$listeners, 'input')).toBe(true);
    done();
  });
  it('default input event should emit model change', async (done) => {
    const elSelectStubVm = wrapper.findAll('el-select-stub').at(0).vm;
    const inputFn = elSelectStubVm.$listeners.input as (val: any) => void;
    inputFn(2);
    const emitted = wrapper.emitted();
    expect(emitted).toEqual({"formChange": [[{"test-val": {"test2": 2}}]]});
    done();
  });
});
