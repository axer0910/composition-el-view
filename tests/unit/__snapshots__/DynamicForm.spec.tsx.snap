// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicForm.tsx should render Dynamic Form 1`] = `
<el-form-stub labelsuffix="" showmessage="true" validateonrulechange="true">
  <div class="form-row">
    <div class="form-col">
      <el-form-item-stub label="测试表单项" inlinemessage="" showmessage="true">
        <el-input-stub type="text" autocomplete="off" validateevent="true" placeholder="测试表单项"></el-input-stub>
      </el-form-item-stub>
    </div>
  </div>
  <div class="form-row">
    <div class="form-col">
      <el-form-item-stub label="测试下拉" inlinemessage="" showmessage="true">
        <el-select-stub autocomplete="off" multiplelimit="0" placeholder="测试表单项" valuekey="value" popperappendtobody="true">
          <el-option-stub value="1" label="下拉1"></el-option-stub>
          <el-option-stub value="2" label="下拉2"></el-option-stub>
        </el-select-stub>
      </el-form-item-stub>
    </div>
  </div>
</el-form-stub>
`;

exports[`DynamicForm.tsx should render jsx 1`] = `
<el-form-stub labelsuffix="" showmessage="true" validateonrulechange="true">
  <div class="form-row">
    <div class="form-col">
      <el-form-item-stub label="jsx下拉选项" inlinemessage="" showmessage="true">
        <el-select-stub autocomplete="off" multiplelimit="0" placeholder="请选择" valuekey="value" popperappendtobody="true">
          <option value="1">test1</option>
          <option value="2">test2</option>
        </el-select-stub>
      </el-form-item-stub>
    </div>
  </div>
</el-form-stub>
`;

exports[`DynamicForm.tsx should update Dynamic Form when prop change 1`] = `
<el-form-stub labelsuffix="" showmessage="true" validateonrulechange="true">
  <div class="form-row">
    <div class="form-col">
      <el-form-item-stub label="新的表单项" inlinemessage="" showmessage="true">
        <el-input-stub type="text" validateevent="true" placeholder="新的表单项"></el-input-stub>
      </el-form-item-stub>
    </div>
  </div>
</el-form-stub>
`;
